cmd/podman/completion/completion.go

```diff
@@ -0,0 +1,85 @@
+package completion
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/containers/libpod/v2/cmd/podman/common"
+	"github.com/containers/libpod/v2/cmd/podman/registry"
+	"github.com/containers/libpod/v2/pkg/domain/entities"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+const (
+	completionDescription = `Generate shell autocompletions.
+  Valid arguments are bash, zsh, and fish.
+  Please refer to the man page to see how you can load these completions.`
+)
+
+var (
+	shells            = []string{"bash", "zsh", "fish"}
+	completionOptions = entities.CompletionOptions{}
+	completionCmd     = &cobra.Command{
+		Use:              fmt.Sprintf("completion [flags] {%s}", strings.Join(shells, ",")),
+		Short:            "Generate shell autocompletions",
+		Long:             completionDescription,
+		TraverseChildren: true,
+		ValidArgsFunction: func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+			return shells, cobra.ShellCompDirectiveNoFileComp
+		},
+		Args: cobra.ExactValidArgs(1),
```

Comment:
This is valid, but it won't do what it is meant to do.  It's not a big deal but I wanted to point it out.

Because you use `ValidArgsFunction` instead of `ValidArgs`, which is the best way in my mind too, so that you can turn off file completion for the shell parameter, the `ExactValidArgs()` function no longer knows that the valid args are, so it will just allow any arg.  But you do catch anything invalid in your code yourself.

This may be something Cobra could improve upon, but I don't have a clear idea yet.